### PR TITLE
Fix issue #9683 "gui, wallet: random abort (segmentation fault)

### PR DESCRIPTION
**Pull request by dooglus**
Patch taken from @ryanofsky's comment https://github.com/bitcoin/bitcoin/issues/9683#issuecomment-448035913.

[MarcoFalke wrote](https://github.com/bitcoin/bitcoin/issues/9683#issuecomment-454066004):
> Mind to submit this patch as a pull request?

So that's what I'm doing.

I was regularly seeing crashes on startup before applying this patch and haven't seen a single crash on startup since applying it almost a month ago.

### CODE REVIEW

**Review by ryanofsky — APPROVED** [review]
utACK ebb0758d14311496e1b1f676c74d5953663e8660. Change should be ok. We never actually narrowed down exactly what was causing the abort to happen or why the change helps, but I have three theories:

1) ~~Missing `setSortRole` call~~ was causing TransactionTableModel::data() to return some unknown field with an inconsistent (non-transitive?) sort order that made Qt crash internally.
2) Not specifying a sort column before calling `setSortingEnabled(true)` was causing a similar problem.
3) Code is still buggy, and this change only masks the problem. Abort is just less likely to happen now because removed `sortByColumn` call means sort only happens once instead of twice in a row.

It would probably be possible to experiment more and narrow down the cause, but I think not necessary. 

Maybe do keep a backup of the crashing wallet around, to help reproduce in case similar bugs happen in the future.

---

EDIT: `setSortRole` call was not actually missing: https://github.com/bitcoin/bitcoin/pull/15203#discussion_r249168229, so this was probably not the cause.

**Comment by dooglus** [comment]
> keep a backup of the crashing wallet around

It's my main Bitcoin wallet that I've been using since 2011. I'll be keeping it around.

File: src/qt/transactionview.cpp [hunk 1]
```diff
@@ -218,6 +218,7 @@ void TransactionView::setModel(WalletModel *_model)
     {
         transactionProxyModel = new TransactionFilterProxy(this);
         transactionProxyModel->setSourceModel(_model->getTransactionTableModel());
+        transactionProxyModel->setSortRole(Qt::EditRole);
```
Comment (hebasto) [inline src/qt/transactionview.cpp]: This line is a duplicate of that one:
https://github.com/bitcoin/bitcoin/blob/9096276e0b2d5b7e19af9a5f3c144ef108ee55e0/src/qt/transactionview.cpp#L225

**Review by hebasto — CHANGES_REQUESTED** [review]

File: src/qt/transactionview.cpp (hunk 1 above)
Comment (ryanofsky) [reply src/qt/transactionview.cpp]: ccv

Comment (ryanofsky) [reply src/qt/transactionview.cpp]: re: https://github.com/bitcoin/bitcoin/pull/15203#discussion_r249168229

> This line is a duplicate of that one:

Good catch. I think it means this line is not actually doing anything and can be removed.

Comment (hebasto) [reply src/qt/transactionview.cpp]: @ryanofsky why does the default `Qt::DisplayRole` does not work in this case?

Comment (laanwj) [reply src/qt/transactionview.cpp]: > but it looks like the intention was to use EditRole to sort based on unformatted values, while letting DisplayRole return formatted values:

Yep.

Comment (hebasto) [reply src/qt/transactionview.cpp]: @ryanofsky @laanwj thanks for clarification.

**Review by ryanofsky — COMMENTED** [review]
It would be nice to remove the extra setSortRole call: https://github.com/bitcoin/bitcoin/pull/15203#discussion_r249168229. But I think it'd be fine to merge this PR with or without that change.

**Comment by dooglus** [comment]
I can do that if you like, but the older hardware I was previously using finally quit working. I'm now on much faster CPU & storage so don't know if I can even reproduce the issue any more.

**Comment by dooglus** [comment]
I did a `git reset HEAD^` and `git push --force`. Was that wrong? I don't see the discussion of the previous commit any more unless I click the comment link I put in the commit message.

**Pull request merged**